File: hw2_files/p3_code.cpp
//insert your code in Part_1/p1_opencv_code.cpp
//edit the file extension and web template to match your programing language

#include <iostream>
#include <ctime>
#include <opencv2/opencv.hpp>
#include "label_components.h"

int main(int argc, char** argv) {

	cv::Mat img_gry, img_bw;
	img_gry = cv::imread("../p3_search.png", 0); // flag = 0 => CV_LOAD_IMAGE_GRAYSCALE

	if (img_gry.empty()) {
		std::cout << "Not a valid image file" << std::endl;
		return -1;
	}

	cv::threshold(img_gry, img_bw, 0.5, 255, cv::THRESH_BINARY);
	cv::Mat compMat = label_components(img_bw);

	int rows = compMat.rows;
	int cols = compMat.cols;

	double _min, max;
	cv::minMaxIdx(compMat, &_min, &max);
	int numObjects = (int)max;
	cv::Mat colors = cv::Mat(1, numObjects, CV_8UC3);
	cv::RNG rng = cv::RNG(std::time(NULL));
	rng.fill(colors, rng.UNIFORM, cv::Scalar(0,0,0), cv::Scalar(255,255,255));
	cv::Mat coloredImg = cv::Mat::zeros(rows, cols, CV_8UC3);
	colors.at<cv::Vec3b>(0,0) = cv::Vec3b(255,255,255);
	for (int row = 0; row < rows; ++row) {
		for (int col = 0; col < cols; ++col) {
			if (int id = compMat.at<ushort>(row, col)) {
				coloredImg.at<cv::Vec3b>(row, col) = colors.at<cv::Vec3b>(0, id-1);
			}
		}
	}

	cv::imwrite("p3_image.png", coloredImg);

	return 0;
}

File: hw2_files/p3_code.m
%insert your code in Part_3/p3_code.cpp
%edit the file extension and web template to match your programing language

I = imread("p3_search.png");
I = im2gray(I);
I = imbinarize(I);

[rows, cols] = size(I);
objectMat = zeros(rows+1,cols+1);
i = 1;
for row = 1:rows
    for col = 1:cols
        if (I(row,col))
            if (objectMat(row,col+1))
                objectMat(row+1,col+1) = objectMat(row,col+1);
                if (objectMat(row+1,col) && objectMat(row+1,col) ~= objectMat(row,col+1))
                    % missing equivalency table addition
                end
            elseif (objectMat(row+1,col))
                objectMat(row+1,col+1) = objectMat(row+1,col);
            else
                objectMat(row+1,col+1) = i;
                i = i + 1;
            end
        end
    end
end
objectMat = objectMat(2:end,2:end);

% Missing equivalency table simplification

% Color each identified object/sqaure
numObjects = max(objectMat,[],"all");
colors = rand(numObjects, 3);
colors(1,:) = [1 1 1]; % set 1st element (outside border) to white
coloredMat = zeros(rows,cols,3);
for row = 1:rows
    for col = 1:cols
        if (objectMat(row,col) > 0)
            coloredMat(row,col,:) = colors(objectMat(row,col),:);
        end
    end
end

montage({I,objectMat,coloredMat})

imwrite(coloredMat, "p3_image.png");